### PR TITLE
Added support for AWS KMS transparent decryption

### PR DESCRIPTION
**Pull request by xjrk58**
- Added support for AWS KMS transparent decryption of the whole configuration json for sensitive data.

### CODE REVIEW

File: src/main/java/org/zalando/baigan/service/aws/S3FileLoader.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        }
+       return candidate;
+    }
+
+    private ByteBuffer decryptValue(final byte[] encryptedBytes) {
+        final DecryptRequest request = new DecryptRequest().withCiphertextBlob(ByteBuffer.wrap(encryptedBytes));
+        short tries = 0;
+        while(true) {
+            try {
+                return kmsClient.decrypt(request).getPlaintext(); // This seems to randomly throw exceptions
```
Comment (lmineiro) [inline src/main/java/org/zalando/baigan/service/aws/S3FileLoader.java]: I understand the comments but would be way more confident if we get down to the real reason

File: src/main/java/org/zalando/baigan/service/aws/S3FileLoader.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+
+/* Provides transparent content decryption of encrypted configuration content using AWS KMS. All configuration values
+ * surrounded by {@value #KMS_START_TAG} and {@value #KMS_END_TAG} are decrypted automatically. The content must
+ * be Base64 encoded and the decrypted content is interpreted as an UTF-8 encoded string. */
+
+public class S3FileLoader {
+
+    private static final String KMS_START_TAG = "awskms{";
```
Comment (lmineiro) [inline src/main/java/org/zalando/baigan/service/aws/S3FileLoader.java]: I many other, related/similar, tools there's a standard of just prefixing with `awk:kms:`
For example, an encrypted key value would contain:
  `aws:kms:fooBar` if the encrypted value was fooBar.
Also seems a bit less complex than having an opening and closing tag.

File: src/main/java/org/zalando/baigan/service/aws/S3FileLoader.java (hunk 1 above)
Comment (xjrk58) [inline src/main/java/org/zalando/baigan/service/aws/S3FileLoader.java]: I've changed the retry logic to retry on `DependencyTimeoutException` and `KMSInternalException` only which should be enough...

File: src/main/java/org/zalando/baigan/service/aws/S3FileLoader.java (hunk 2 above)
Comment (xjrk58) [inline src/main/java/org/zalando/baigan/service/aws/S3FileLoader.java]: Aligned with the standard.

File: src/main/java/org/zalando/baigan/service/aws/S3FileLoader.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        final Optional<byte[]> encryptedValue = getEncryptedValue(candidate);
+        if (encryptedValue.isPresent()) {
+            return new String(toByteArray(decryptValue(encryptedValue.get())), StandardCharsets.UTF_8);
+        }
+       return candidate;
+    }
+
+    private ByteBuffer decryptValue(final byte[] encryptedBytes) {
+        final DecryptRequest request = new DecryptRequest().withCiphertextBlob(ByteBuffer.wrap(encryptedBytes));
+        short tries = 0;
```
Comment (lmineiro) [inline src/main/java/org/zalando/baigan/service/aws/S3FileLoader.java]: I'd be way happier if we used something like https://github.com/rholder/guava-retrying#exponential-backoff

I'd set the maximum the the current 10sec and the stop strategy after 5 attempts

**Comment by lmineiro** [comment]
lgtm
@amjedonline ?

**Comment by lmineiro** [comment]
I'll wait for @amjedonline's feedback while I draft a new release

File: src/main/java/org/zalando/baigan/service/S3ConfigurationRepository.java [hunk 4]
```diff
@@ -61,7 +67,7 @@ public S3ConfigurationRepository(@Nonnull final String bucketName, @Nonnull fina
         this.key = key;
         this.refreshInterval = refreshInterval;
 
-        s3Client = new AmazonS3Client();
+        s3Loader = new S3FileLoader(bucketName, key);
         loadConfigurations();
 
         if (refreshInterval > 0) {
```
Comment (amjedonline) [inline src/main/java/org/zalando/baigan/service/S3ConfigurationRepository.java]: Redundant check ?

**Comment by amjedonline** [comment]
I have merged the PR.
There are a couple of thoughts though.
1. Can we have a way to forego the load request if an earlier call to load the content is still in waiting/backoff process ?
2. @lmineiro like to see [Hystrix](https://github.com/Netflix/Hystrix) integration in your new release  :)
3.  Refactor Github, Etcd Repositories to reuse the above resilience/retry method.. or shall I create a new Issue for that ?

**Comment by lmineiro** [comment]
Regarding 1, I've thought about it and it won't come to that. The reload happens every 60secs and the retry strategy should fail before that. I personally never thought about so large configurations that they could end up using the entire 60sec slot to download and process. A locking mechanism would prevent this.
Maybe a separate issue.

Regarding 2 and 3. Why not.

**Comment by xjrk58** [comment]
Quick note regarding 1 - KMS allows to encrypt blocks only up to 4kB, so download time shouldn't be a problem at all. The problem could be only if KMS becomes temporarily unavailable and that's when the retry strategy should kick in.

**Comment by lmineiro** [comment]
I wasn't personally worried about the KMS decryption alone. It's about fetching the JSON + processing the entire configuration (including multiple decryption requests). This is triggered by a background thread. In case of slow networks, retransmission, retries or whatever reasons, both steps can actually take "some time".
@amjedonline's concerns where about making sure that whatever the time, the background thread wouldn't start triggering multiple concurrent configuration updates.
I'm very confident that for this repository (S3) we did some efforts to prevent it. We used the `ScheduledThreadPoolExecutor` with a size of 1. That means that if the only thread in the pool is still busy, we won't trigger another background reload until that one is done. It may throw an exception though.

**Pull request merged**